tensorflow_probability/python/mcmc/replica_exchange_mc.py

```diff
@@ -0,0 +1,365 @@
+# Copyright 2018 The TensorFlow Probability Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""Replica Exchange Monte Carlo Transition Kernel."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+import tensorflow as tf
+
+from tensorflow_probability.python.mcmc import kernel as kernel_base
+from tensorflow_probability.python.mcmc import util as mcmc_util
+from tensorflow.python.ops.distributions import util as distributions_util
+
+
+__all__ = [
+    'ReplicaExchangeMC',
+    'default_exchange_proposed_fn',
+]
+
+
+ReplicaExchangeMCKernelResults = collections.namedtuple(
+    'ReplicaExchangeMCKernelResults',
+    [
+        'replica_results',
+        'replica_states',
+        'next_replica_idx',
+        'exchange_proposed',
+        'exchange_proposed_n',
+    ])
+
+
+def default_exchange_proposed_fn(freq):
+  """Default function for `exchange_proposed_fn` of `kernel`.
+  Depending on the probability of `freq`, decide whether to propose combinations
+  of replica for exchange.
+  When exchanging, create combinations of adjacent replicas from 0 or 1 index.
+  """
+  def default_exchange_proposed_fn_(n_replica, seed=None):
+    seed = distributions_util.gen_new_seed(seed, "default_exchange_proposed_fn")
+    u = tf.random_uniform([], seed=seed)
+    exist = u > freq
+
+    seed = distributions_util.gen_new_seed(seed, "default_exchange_proposed_fn")
+    i = tf.random_uniform([], minval=0, maxval=2, dtype=tf.int32, seed=seed)
+    a = tf.range(i, n_replica - 1, 2)
+    exchange_proposed = tf.transpose(tf.concat(([a], [a + 1]), axis=0))
+    exchange_proposed_n = tf.to_int32(tf.floor(
+        (tf.to_int32(n_replica) - i) / 2))
+
+    exchange_proposed = tf.cond(
+        exist, lambda: tf.to_int32([]), lambda: exchange_proposed)
+    exchange_proposed_n = tf.cond(exist, lambda: 0, lambda: exchange_proposed_n)
```

Comment:
For computational reasons its usually better if we can avoid graph control flow.  How about:

    tf.where(exist, tf.zeros_like(exchange_proposed_n), exchange_proposed_n)

I dont see an easy way to rewrite the first use of `tf.cond`, but maybe you do?

Reply:
Thanks.
I'll use slicing.
```
exchange_proposed_n = tf.where(exist, tf.zeros_like(exchange_proposed_n),
                                   exchange_proposed_n)
exchange_proposed = exchange_proposed[:exchange_proposed_n]
```

Reply:
When I use `sample_chain`, `exchange_proposed` whose shape is not invariant cause an error. Therefore, I rewrote this function so that `exchange_proposed` has invariant shape.